stl/inc/xstring

```diff
@@ -2722,35 +2722,62 @@ private:
         }
 
         _Tidy_deallocate_guard<basic_string> _Guard{this};
-        for (; _First != _Last; ++_First) {
-            if constexpr (!is_same_v<_Size, size_type>) {
-                if (_My_data._Mysize == _My_data._Myres) { // Need to grow
-                    if (_My_data._Mysize == max_size()) {
-                        _Xlen_string(); // result too long
-                    }
 
-                    _Elem* const _Old_ptr   = _My_data._Myptr();
-                    size_type _New_capacity = _Calculate_growth(_My_data._Mysize + 1);
-                    const pointer _New_ptr  = _Allocate_for_capacity(_Al, _New_capacity); // throws
+        constexpr bool _Can_construct_by_memcpy = _Is_specialization_v<_Traits, char_traits> && _Is_EcharT<_Elem>
+                                               && _Iterator_is_contiguous<_Iter> && !_Iterator_is_volatile<_Iter>
+                                               && is_integral_v<_Iter_value_t<_Iter>>
+                                               && sizeof(_Iter_value_t<_Iter>) == sizeof(_Elem);

```

Comment:
Could we reduce complexity (and simplify maintenance) by checking only `_Is_specialization_v<_Traits, char_traits> && _Is_EcharT<_Elem> && is_same_v<_Size, size_type>` here and calling `copy_n`? It already does the work to determine when it can use `memcpy` so we don't have to.

Reply:
I'm thinking:
```c++
        _Tidy_deallocate_guard<basic_string> _Guard{this};

        constexpr bool _Can_construct_by_copy =
            _Is_specialization_v<_Traits, char_traits> && _Is_EcharT<_Elem> && is_same_v<_Size, size_type>;

        if constexpr (_Can_construct_by_copy) {
            const auto _Data = _My_data._Myptr();

            if constexpr (_Iterator_is_contiguous<_Iter>) {
                _STD _Copy_n_unchecked4(_STD _To_address(_First), _Count, _Data);
            } else {
                _STD _Copy_n_unchecked4(_STD move(_First), _Count, _Data);
            }

            _My_data._Mysize = _Count;
            _Data[_Count]    = _Elem();
        } else {
```
which nicely restores the warning for initializing `std::string` from a sequence of `unsigned char`.

Reply:
`_Copy_n_unchecked4` works, but it eventually calls `memmove`. Not sure whether this is wanted.

Reply:
There is no perf difference between `memmove` and `memcpy` for us - they are literally the same implementation.

(Of course, we must still be careful to call `memmove` instead of `memcpy` when overlapping is possible.)

Reply:
OK. Since `_Copy_n_unchecked4` dispatches itself, I'm simply calling`_STD _Copy_n_unchecked4(_STD move(_First), _Count, _Data);`.